File: packages/remote-protocol/src/review-publication-attribution.ts
import { Buffer } from "node:buffer";
import { createHash, createPublicKey, verify } from "node:crypto";
import {
  canonicalize,
  fail,
  isCanonicalTimestamp,
  isDigest,
  isIdentifier,
  ok,
  ownData,
  plainActor,
  toPlainRecord,
  type Actor,
  type PlainRecord,
  type PlainValue,
  type SchemaMeta,
  type ValidationIssue,
  type ValidationResult,
} from "@getsimpledirect/vinci-contracts";
import { decodeCanonicalBase64Url } from "@getsimpledirect/vinci-device-auth";
import {
  validateSessionBindingRef,
  type SessionBindingRef,
} from "./binding-ref.ts";
import { parseStrictSignedJson } from "./strict-json.ts";
import {
  issue,
  prefixIssues,
  rejectUnknownFields,
  validateId,
} from "./wire-validation.ts";

export const REVIEW_PUBLICATION_PURPOSE = "guard_review.publish" as const;
export const REVIEW_PUBLICATION_AUDIENCE = "vinci-acceptance" as const;
export const REVIEW_PUBLICATION_VERDICTS = ["GO", "BLOCK"] as const;
export const MAX_REVIEW_PUBLICATION_ATTRIBUTION_LIFETIME_MS = 10 * 60 * 1_000;

export type ReviewPublicationVerdict = (typeof REVIEW_PUBLICATION_VERDICTS)[number];

export type ReviewPublicationGitHubSubject = {
  readonly provider: "github";
  /** Stable GitHub GraphQL repository node id; owner/name is never identity. */
  readonly repositoryNodeId: string;
  readonly pullRequestNumber: number;
  /** Exact lowercase, full Git object ids. Branch names and abbreviations are forbidden. */
  readonly headSha: string;
  readonly baseSha: string;
  readonly headTreeSha: string;
};

/**
 * A VGC-signed transport assertion for one guard-review publication attempt.
 *
 * This value carries no authority by itself. Runtime VGC MUST derive `actor`
 * and every member of `binding` from authenticated server-side state, prove
 * verifier independence there, and re-resolve the GitHub subject before it
 * signs. Vinci Acceptance MUST independently authenticate its caller, resolve
 * a currently usable platform-issuer key, verify this signature and binding,
 * and re-resolve the repository/PR snapshot. A caller-authored value that
 * merely validates against this transport schema is not an identity, session,
 * tenant, reviewer-independence, or publication authorization proof.
 */
export type ReviewPublicationAttribution = {
  readonly schemaVersion: 1;
  readonly purpose: typeof REVIEW_PUBLICATION_PURPOSE;
  readonly audience: typeof REVIEW_PUBLICATION_AUDIENCE;
  readonly actor: Actor;
  readonly binding: SessionBindingRef;
  readonly subject: ReviewPublicationGitHubSubject;
  readonly verdict: ReviewPublicationVerdict;
  readonly recordSetDigest: string;
  readonly idempotencyKey: string;
  readonly issuedAt: string;
  readonly expiresAt: string;
  readonly issuerKeyId: string;
  readonly signature: {
    readonly alg: "Ed25519";
    /** Canonical unpadded base64url encoding of exactly 64 bytes. */
    readonly value: string;
  };
};

export type ReviewPublicationReference = {
  readonly reviewId: string;
  readonly publicationDigest: string;
};

const ATTRIBUTION_FIELDS = [
  "schemaVersion",
  "purpose",
  "audience",
  "actor",
  "binding",
  "subject",
  "verdict",
  "recordSetDigest",
  "idempotencyKey",
  "issuedAt",
  "expiresAt",
  "issuerKeyId",
  "signature",
] as const;
const SUBJECT_FIELDS = [
  "provider",
  "repositoryNodeId",
  "pullRequestNumber",
  "headSha",
  "baseSha",
  "headTreeSha",
] as const;
const FULL_GIT_SHA = /^[0-9a-f]{40}$/;
const OPAQUE_GITHUB_NODE_ID = /^[\x21-\x7e]{1,255}$/;
const REVIEW_ID = /^grv_[A-Za-z0-9][A-Za-z0-9._:-]{0,123}$/;
const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

/** Python datetime and this wire contract share the ISO-8601 year 0001..9999 domain. */
function isReviewPublicationTimestamp(value: unknown): value is string {
  return isCanonicalTimestamp(value) && !(value as string).startsWith("0000-");
}

function assertSigningTimePair(attribution: ReviewPublicationAttribution): void {
  if (!isReviewPublicationTimestamp(attribution.issuedAt)
      || !isReviewPublicationTimestamp(attribution.expiresAt)) {
    throw new Error("review-publication signing timestamps must be canonical years 0001 through 9999");
  }
  const lifetime = Date.parse(attribution.expiresAt) - Date.parse(attribution.issuedAt);
  if (lifetime <= 0 || lifetime > MAX_REVIEW_PUBLICATION_ATTRIBUTION_LIFETIME_MS) {
    throw new Error("review-publication signing timestamps must be ordered within the ten-minute lifetime");
  }
}

function recordAt(
  value: unknown,
  path: string,
  issues: ValidationIssue[],
): PlainRecord | undefined {
  const plain = toPlainRecord(value);
  if (!plain.ok) {
    prefixIssues(path, plain.issues, issues);
    return undefined;
  }
  return plain.value;
}

function escapePointer(value: string): string {
  return value.replaceAll("~", "~0").replaceAll("/", "~1");
}

function validateUnicode(
  value: PlainValue,
  path: string,
  issues: ValidationIssue[],
): void {
  if (typeof value === "string") {
    for (let i = 0; i < value.length; i += 1) {
      const code = value.charCodeAt(i);
      if (code >= 0xd800 && code <= 0xdbff) {
        const next = value.charCodeAt(i + 1);
        if (next < 0xdc00 || next > 0xdfff) {
          issues.push(issue(path, "invalid_unicode", "strings must contain only Unicode scalar values"));
          return;
        }
        i += 1;
      } else if (code >= 0xdc00 && code <= 0xdfff) {
        issues.push(issue(path, "invalid_unicode", "strings must contain only Unicode scalar values"));
        return;
      }
    }
    if (value.normalize("NFC") !== value) {
      issues.push(issue(path, "non_canonical_unicode", "strings must use Unicode NFC normalization"));
    }
    return;
  }
  if (Array.isArray(value)) {
    value.forEach((entry, index) => validateUnicode(entry, `${path}/${index}`, issues));
    return;
  }
  if (value !== null && typeof value === "object") {
    for (const [key, entry] of Object.entries(value)) {
      validateUnicode(key, `${path}/${escapePointer(key)}`, issues);
      validateUnicode(entry, `${path}/${escapePointer(key)}`, issues);
    }
  }
}

function validateSubject(
  value: unknown,
  originalValue: unknown,
  issues: ValidationIssue[],
): ReviewPublicationGitHubSubject | undefined {
  const subject = recordAt(value, "/subject", issues);
  if (subject === undefined) return undefined;
  rejectUnknownFields(subject, SUBJECT_FIELDS, "/subject", issues);
  if (subject.provider !== "github") {
    issues.push(issue("/subject/provider", "invalid_provider", "provider must be github"));
  }
  if (typeof subject.repositoryNodeId !== "string" || !OPAQUE_GITHUB_NODE_ID.test(subject.repositoryNodeId)) {
    issues.push(issue("/subject/repositoryNodeId", "invalid_github_node_id", "repositoryNodeId must be a printable ASCII GitHub node id"));
  }
  const original = ownData(originalValue, "pullRequestNumber");
  if (
    !Number.isSafeInteger(subject.pullRequestNumber)
    || (subject.pullRequestNumber as number) < 1
    || Object.is(original, -0)
  ) {
    issues.push(issue("/subject/pullRequestNumber", "invalid_pull_request_number", "pullRequestNumber must be a positive safe integer"));
  }
  for (const field of ["headSha", "baseSha", "headTreeSha"] as const) {
    if (typeof subject[field] !== "string" || !FULL_GIT_SHA.test(subject[field])) {
      issues.push(issue(`/subject/${field}`, "invalid_git_sha", `${field} must be an exact lowercase 40-hex Git object id`));
    }
  }
  return subject as unknown as ReviewPublicationGitHubSubject;
}

function validateSignature(
  value: unknown,
  issues: ValidationIssue[],
): ReviewPublicationAttribution["signature"] | undefined {
  const signature = recordAt(value, "/signature", issues);
  if (signature === undefined) return undefined;
  rejectUnknownFields(signature, ["alg", "value"], "/signature", issues);
  if (signature.alg !== "Ed25519") {
    issues.push(issue("/signature/alg", "invalid_signature_algorithm", "only Ed25519 is supported"));
  }
  if (decodeCanonicalBase64Url(signature.value)?.byteLength !== 64) {
    issues.push(issue("/signature/value", "invalid_signature_value", "signature.value must be canonical unpadded base64url encoding of exactly 64 bytes"));
  }
  return signature as unknown as ReviewPublicationAttribution["signature"];
}

/**
 * Strict transport validation at an explicit trusted clock instant.
 *
 * Passing does not verify the signature, key role/status, caller, tenancy,
 * repository state, session binding, reviewer identity, or independence.
 */
export function validateReviewPublicationAttribution(
  input: unknown,
  now: string,
): ValidationResult<ReviewPublicationAttribution> {
  const plain = toPlainRecord(input);
  if (!plain.ok) return plain;
  const record = plain.value;
  const issues: ValidationIssue[] = [];
  validateUnicode(record, "", issues);
  rejectUnknownFields(record, ATTRIBUTION_FIELDS, "", issues);

  if (!isReviewPublicationTimestamp(now)) {
    issues.push(issue("/", "invalid_validation_time", "expiry validation requires a canonical UTC timestamp"));
  }
  if (record.schemaVersion !== REVIEW_PUBLICATION_ATTRIBUTION_SCHEMA_META.version) {
    issues.push(issue("/schemaVersion", "schema_version_mismatch", "unsupported review-publication attribution schema version"));
  }
  if (record.purpose !== REVIEW_PUBLICATION_PURPOSE) {
    issues.push(issue("/purpose", "invalid_purpose", `purpose must be ${REVIEW_PUBLICATION_PURPOSE}`));
  }
  if (record.audience !== REVIEW_PUBLICATION_AUDIENCE) {
    issues.push(issue("/audience", "invalid_audience", `audience must be ${REVIEW_PUBLICATION_AUDIENCE}`));
  }

  const actor = typeof record.actor === "object" && record.actor !== null && !Array.isArray(record.actor)
    ? plainActor(record.actor as Readonly<Record<string, unknown>>)
    : null;
  if (actor === null) {
    issues.push(issue("/actor", "invalid_actor", "actor must be one exact member of the central Actor union"));
  } else {
    const identityFields = actor.kind === "user"
      ? ["userId", ...(actor.deviceId === undefined ? [] : ["deviceId"])] as const
      : actor.kind === "worker"
        ? ["workerId"] as const
        : actor.kind === "policy"
          ? ["policyId"] as const
          : actor.kind === "system"
            ? ["component"] as const
            : ["verifierId"] as const;
    for (const field of identityFields) {
      if (!isIdentifier(actor[field])) {
        issues.push(issue(`/actor/${field}`, "invalid_id", `${field} must be an identifier`));
      }
    }
    if (actor.kind === "policy" && !Number.isSafeInteger(actor.policyVersion)) {
      issues.push(issue("/actor/policyVersion", "unsafe_integer", "policyVersion must be a positive safe integer"));
    }
  }

  const binding = validateSessionBindingRef(record.binding);
  if (!binding.ok) prefixIssues("/binding", binding.issues, issues);
  const subject = validateSubject(
    record.subject,
    ownData(input, "subject"),
    issues,
  );

  if (!(REVIEW_PUBLICATION_VERDICTS as readonly unknown[]).includes(record.verdict)) {
    issues.push(issue("/verdict", "invalid_verdict", "verdict must be GO or BLOCK"));
  }
  if (!isDigest(record.recordSetDigest)) {
    issues.push(issue("/recordSetDigest", "invalid_digest", "recordSetDigest must be lowercase SHA-256"));
  }
  for (const field of ["idempotencyKey", "issuerKeyId"] as const) {
    validateId(record[field], `/${field}`, issues);
  }
  if (!isReviewPublicationTimestamp(record.issuedAt)) {
    issues.push(issue("/issuedAt", "invalid_timestamp", "issuedAt must be a canonical UTC timestamp in years 0001 through 9999"));
  }
  if (!isReviewPublicationTimestamp(record.expiresAt)) {
    issues.push(issue("/expiresAt", "invalid_timestamp", "expiresAt must be a canonical UTC timestamp in years 0001 through 9999"));
  }
  if (isReviewPublicationTimestamp(record.issuedAt) && isReviewPublicationTimestamp(record.expiresAt)) {
    const issuedMs = Date.parse(record.issuedAt);
    const expiresMs = Date.parse(record.expiresAt);
    if (expiresMs <= issuedMs) {
      issues.push(issue("/expiresAt", "invalid_time_order", "expiresAt must be strictly later than issuedAt"));
    } else if (expiresMs - issuedMs > MAX_REVIEW_PUBLICATION_ATTRIBUTION_LIFETIME_MS) {
      issues.push(issue("/expiresAt", "lifetime_exceeded", "a review-publication attribution may live for at most 10 minutes"));
    }
    if (isReviewPublicationTimestamp(now) && expiresMs <= Date.parse(now)) {
      issues.push(issue("/expiresAt", "expired", "the review-publication attribution has expired"));
    }
  }
  const signature = validateSignature(record.signature, issues);

  if (
    issues.length > 0
    || actor === null
    || !binding.ok
    || subject === undefined
    || signature === undefined
  ) {
    return fail(issues);
  }
  return ok(record as unknown as ReviewPublicationAttribution);
}

/** Rejects duplicate fields, invalid UTF-8/Unicode and unsafe JSON numbers before shape validation. */
export function parseReviewPublicationAttributionJson(
  input: string | Uint8Array,
  now: string,
): ValidationResult<ReviewPublicationAttribution> {
  const parsed = parseStrictSignedJson(input);
  if (!parsed.ok) return parsed;
  return validateReviewPublicationAttribution(parsed.value, now);
}

/** Canonical UTF-8 bytes signed by VGC; only signature.value is excluded. */
export function reviewPublicationAttributionSigningPayload(
  attribution: ReviewPublicationAttribution,
): Uint8Array {
  assertSigningTimePair(attribution);
  return new TextEncoder().encode(canonicalize({
    schemaVersion: attribution.schemaVersion,
    purpose: attribution.purpose,
    audience: attribution.audience,
    actor: attribution.actor,
    binding: attribution.binding,
    subject: attribution.subject,
    verdict: attribution.verdict,
    recordSetDigest: attribution.recordSetDigest,
    idempotencyKey: attribution.idempotencyKey,
    issuedAt: attribution.issuedAt,
    expiresAt: attribution.expiresAt,
    issuerKeyId: attribution.issuerKeyId,
    signature: { alg: attribution.signature.alg },
  }));
}

/** SHA-256 of the exact canonical signing bytes, excluding signature.value. */
export function reviewPublicationAttributionDigest(
  attribution: ReviewPublicationAttribution,
): string {
  return createHash("sha256")
    .update(reviewPublicationAttributionSigningPayload(attribution))
    .digest("hex");
}

/**
 * Cryptographic check only, against one raw Ed25519 public key.
 *
 * A true result does not establish that the key is a usable VGC platform
 * issuer or that the claims match server state; Acceptance must check those
 * facts independently.
 */
export function verifyReviewPublicationAttributionSignature(
  attribution: ReviewPublicationAttribution,
  publicKey: string,
): boolean {
  try {
    const rawKey = decodeCanonicalBase64Url(publicKey);
    const signature = decodeCanonicalBase64Url(attribution.signature.value);
    if (rawKey?.byteLength !== 32 || signature?.byteLength !== 64) return false;
    const key = createPublicKey({
      key: Buffer.concat([ED25519_SPKI_PREFIX, Buffer.from(rawKey)]),
      format: "der",
      type: "spki",
    });
    return verify(
      null,
      reviewPublicationAttributionSigningPayload(attribution),
      key,
      Buffer.from(signature),
    );
  } catch {
    return false;
  }
}

export function validateReviewPublicationReference(
  input: unknown,
): ValidationResult<ReviewPublicationReference> {
  if (typeof input !== "string") {
    return fail([issue("/", "invalid_review_reference", "expected a compact guard-review reference string")]);
  }
  const separator = "@sha256:";
  const separatorAt = input.indexOf(separator);
  if (separatorAt < 1 || separatorAt !== input.lastIndexOf(separator)) {
    return fail([issue("/", "invalid_review_reference", "expected grv_<id>@sha256:<lowercase publication digest>")]);
  }
  const reviewId = input.slice(0, separatorAt);
  const publicationDigest = input.slice(separatorAt + separator.length);
  if (!REVIEW_ID.test(reviewId) || !isDigest(publicationDigest)) {
    return fail([issue("/", "invalid_review_reference", "review id or publication digest is malformed")]);
  }
  return ok(Object.freeze({ reviewId, publicationDigest }));
}

export function formatReviewPublicationReference(
  reviewId: unknown,
  publicationDigest: unknown,
): ValidationResult<string> {
  if (typeof reviewId !== "string" || typeof publicationDigest !== "string") {
    return fail([issue("/", "invalid_review_reference", "review id and publication digest must be strings")]);
  }
  const candidate = `${reviewId}@sha256:${publicationDigest}`;
  const checked = validateReviewPublicationReference(candidate);
  return checked.ok ? ok(candidate) : checked;
}

export const REVIEW_PUBLICATION_ATTRIBUTION_SCHEMA_META = {
  id: "vinci.review-publication-attribution",
  version: 1,
  compatibility: "frozen",
  unknownFields: "reject",
  malformedData: "fail-closed",
  migration: "none",
} as const satisfies SchemaMeta;

File: packages/remote-protocol/python/review_publication_attribution.py
"""Strict Python implementation of ReviewPublicationAttribution v1.

This module mirrors the TypeScript transport contract for cross-language
fixtures. Validation never grants publication authority: VGC must derive actor
and session binding server-side, and Acceptance must authenticate, resolve the
issuer key and role/status, verify the signature, and re-resolve the PR.
"""

import base64
import binascii
import hashlib
import json
import re
import unicodedata
from datetime import datetime


PURPOSE = "guard_review.publish"
AUDIENCE = "vinci-acceptance"
MAX_LIFETIME_MS = 600_000
SAFE_INTEGER = 9_007_199_254_740_991

ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$")
REVIEW_ID = re.compile(r"^grv_[A-Za-z0-9][A-Za-z0-9._:-]{0,123}$")
DIGEST = re.compile(r"^[0-9a-f]{64}$")
GIT_SHA = re.compile(r"^[0-9a-f]{40}$")
NODE_ID = re.compile(r"^[\x21-\x7e]{1,255}$")
TIMESTAMP = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$")
BASE64URL = re.compile(r"^[A-Za-z0-9_-]+$")

ATTRIBUTION_FIELDS = {
    "schemaVersion", "purpose", "audience", "actor", "binding", "subject",
    "verdict", "recordSetDigest", "idempotencyKey", "issuedAt", "expiresAt",
    "issuerKeyId", "signature",
}
BINDING_FIELDS = {
    "protocolVersion", "organizationId", "workspaceId", "runId", "sessionId",
}
SUBJECT_FIELDS = {
    "provider", "repositoryNodeId", "pullRequestNumber", "headSha", "baseSha",
    "headTreeSha",
}
SIGNATURE_FIELDS = {"alg", "value"}
ACTOR_RULES = {
    "user": ({"kind", "userId", "deviceId"}, {"userId"}),
    "worker": ({"kind", "workerId"}, {"workerId"}),
    "policy": ({"kind", "policyId", "policyVersion"}, {"policyId", "policyVersion"}),
    "system": ({"kind", "component"}, {"component"}),
    "verifier": ({"kind", "verifierId", "independent"}, {"verifierId", "independent"}),
}


class ContractError(ValueError):
    """Fail-closed wire validation error with a stable code."""

    def __init__(self, code, message):
        super().__init__(message)
        self.code = code


def _error(code, message):
    raise ContractError(code, message)


def _pairs(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            _error("duplicate_field", "duplicate object member names are forbidden")
        result[key] = value
    return result


def _integer(token):
    value = int(token)
    if token == "-0" or abs(value) > SAFE_INTEGER:
        _error("unsafe_integer", "JSON integers must be safe integers other than -0")
    return value


def _float(_token):
    _error("ambiguous_number", "signed review attribution JSON permits only safe integers")


def parse_strict_json(source):
    """Parse UTF-8 JSON without duplicate keys, floats, unsafe integers or surrogates."""
    if isinstance(source, bytes):
        try:
            source = source.decode("utf-8", "strict")
        except UnicodeDecodeError as exc:
            raise ContractError("invalid_utf8", "signed JSON bytes must be valid UTF-8") from exc
    if not isinstance(source, str):
        _error("invalid_json_input", "expected a JSON string or UTF-8 bytes")
    try:
        value = json.loads(
            source,
            object_pairs_hook=_pairs,
            parse_int=_integer,
            parse_float=_float,
            parse_constant=lambda _value: _error("invalid_json", "non-finite JSON number"),
        )
    except ContractError:
        raise
    except (json.JSONDecodeError, UnicodeError) as exc:
        raise ContractError("invalid_json", "signed JSON could not be parsed") from exc
    _validate_unicode(value)
    return value


def _validate_unicode(value):
    if isinstance(value, str):
        for char in value:
            if 0xD800 <= ord(char) <= 0xDFFF:
                _error("invalid_unicode", "strings must contain only Unicode scalar values")
        if unicodedata.normalize("NFC", value) != value:
            _error("non_canonical_unicode", "strings must use Unicode NFC normalization")
    elif isinstance(value, list):
        for item in value:
            _validate_unicode(item)
    elif isinstance(value, dict):
        for key, item in value.items():
            _validate_unicode(key)
            _validate_unicode(item)


def _record(value, label):
    if type(value) is not dict:
        _error("invalid_record", f"{label} must be a plain object")
    return value


def _exact_fields(value, expected, label):
    keys = set(value)
    if keys != expected:
        _error("invalid_fields", f"{label} must carry exactly its declared fields")


def _identifier(value, label):
    if not isinstance(value, str) or ID.fullmatch(value) is None:
        _error("invalid_id", f"{label} must be an identifier")


def _timestamp(value, label):
    if not isinstance(value, str) or TIMESTAMP.fullmatch(value) is None:
        _error("invalid_timestamp", f"{label} must be a canonical UTC timestamp")
    try:
        return datetime.strptime(value, "%Y-%m-%dT%H:%M:%S.%fZ")
    except ValueError as exc:
        raise ContractError("invalid_timestamp", f"{label} must exist") from exc


def _canonical_base64url(value, length, label):
    if not isinstance(value, str) or BASE64URL.fullmatch(value) is None:
        _error("invalid_base64url", f"{label} must be canonical unpadded base64url")
    try:
        decoded = base64.urlsafe_b64decode(value + "=" * ((4 - len(value) % 4) % 4))
    except (ValueError, binascii.Error) as exc:
        raise ContractError("invalid_base64url", f"{label} is malformed") from exc
    encoded = base64.urlsafe_b64encode(decoded).decode("ascii").rstrip("=")
    if encoded != value or len(decoded) != length:
        _error("invalid_base64url", f"{label} has a non-canonical encoding or wrong length")
    return decoded


def _validate_actor(value):
    actor = _record(value, "actor")
    kind = actor.get("kind")
    if kind not in ACTOR_RULES:
        _error("invalid_actor", "actor must be one member of the central Actor union")
    permitted, required = ACTOR_RULES[kind]
    if not required.issubset(actor) or not set(actor).issubset(permitted):
        _error("invalid_actor", "actor fields do not match its kind")
    for field in required:
        current = actor[field]
        if field == "independent":
            if type(current) is not bool:
                _error("invalid_actor", "independent must be boolean")
        elif field == "policyVersion":
            if type(current) is not int or current < 1 or current > SAFE_INTEGER:
                _error("unsafe_integer", "policyVersion must be a positive safe integer")
        elif not isinstance(current, str) or not current.strip():
            _error("invalid_actor", f"{field} must be non-blank text")
    if "deviceId" in actor and (not isinstance(actor["deviceId"], str) or not actor["deviceId"].strip()):
        _error("invalid_actor", "deviceId must be non-blank text")
    identity_fields = {
        "user": ("userId", "deviceId"),
        "worker": ("workerId",),
        "policy": ("policyId",),
        "system": ("component",),
        "verifier": ("verifierId",),
    }[kind]
    for field in identity_fields:
        if field in actor:
            _identifier(actor[field], field)


def _validate_binding(value):
    binding = _record(value, "binding")
    _exact_fields(binding, BINDING_FIELDS, "binding")
    if binding["protocolVersion"] != 1 or type(binding["protocolVersion"]) is not int:
        _error("protocol_version_mismatch", "binding protocolVersion must be 1")
    if binding["organizationId"] is not None:
        _identifier(binding["organizationId"], "organizationId")
    for field in ("workspaceId", "runId", "sessionId"):
        _identifier(binding[field], field)


def _validate_subject(value):
    subject = _record(value, "subject")
    _exact_fields(subject, SUBJECT_FIELDS, "subject")
    if subject["provider"] != "github":
        _error("invalid_provider", "provider must be github")
    if not isinstance(subject["repositoryNodeId"], str) or NODE_ID.fullmatch(subject["repositoryNodeId"]) is None:
        _error("invalid_github_node_id", "repositoryNodeId is malformed")
    number = subject["pullRequestNumber"]
    if type(number) is not int or number < 1 or number > SAFE_INTEGER:
        _error("invalid_pull_request_number", "pullRequestNumber must be a positive safe integer")
    for field in ("headSha", "baseSha", "headTreeSha"):
        if not isinstance(subject[field], str) or GIT_SHA.fullmatch(subject[field]) is None:
            _error("invalid_git_sha", f"{field} must be exact lowercase 40-hex")


def validate_attribution(value, now):
    """Return a detached validated snapshot, or raise ContractError."""
    _validate_unicode(value)
    record = _record(value, "attribution")
    _exact_fields(record, ATTRIBUTION_FIELDS, "attribution")
    if record["schemaVersion"] != 1 or type(record["schemaVersion"]) is not int:
        _error("schema_version_mismatch", "schemaVersion must be 1")
    if record["purpose"] != PURPOSE:
        _error("invalid_purpose", f"purpose must be {PURPOSE}")
    if record["audience"] != AUDIENCE:
        _error("invalid_audience", f"audience must be {AUDIENCE}")
    _validate_actor(record["actor"])
    _validate_binding(record["binding"])
    _validate_subject(record["subject"])
    if record["verdict"] not in ("GO", "BLOCK"):
        _error("invalid_verdict", "verdict must be GO or BLOCK")
    if not isinstance(record["recordSetDigest"], str) or DIGEST.fullmatch(record["recordSetDigest"]) is None:
        _error("invalid_digest", "recordSetDigest must be lowercase SHA-256")
    _identifier(record["idempotencyKey"], "idempotencyKey")
    _identifier(record["issuerKeyId"], "issuerKeyId")
    issued = _timestamp(record["issuedAt"], "issuedAt")
    expires = _timestamp(record["expiresAt"], "expiresAt")
    current = _timestamp(now, "now")
    lifetime_ms = int((expires - issued).total_seconds() * 1000)
    if lifetime_ms <= 0:
        _error("invalid_time_order", "expiresAt must be strictly later than issuedAt")
    if lifetime_ms > MAX_LIFETIME_MS:
        _error("lifetime_exceeded", "attribution lifetime must not exceed ten minutes")
    if expires <= current:
        _error("expired", "attribution has expired")
    signature = _record(record["signature"], "signature")
    _exact_fields(signature, SIGNATURE_FIELDS, "signature")
    if signature["alg"] != "Ed25519":
        _error("invalid_signature_algorithm", "only Ed25519 is supported")
    _canonical_base64url(signature["value"], 64, "signature.value")
    # JSON contains only the types allowed above, so this gives a detached copy.
    return json.loads(json.dumps(record, ensure_ascii=False))


def parse_attribution_json(source, now):
    return validate_attribution(parse_strict_json(source), now)


def canonical_signing_bytes(attribution):
    """Return the exact bytes signed by VGC; signature.value is excluded."""
    covered = {
        "schemaVersion": attribution["schemaVersion"],
        "purpose": attribution["purpose"],
        "audience": attribution["audience"],
        "actor": attribution["actor"],
        "binding": attribution["binding"],
        "subject": attribution["subject"],
        "verdict": attribution["verdict"],
        "recordSetDigest": attribution["recordSetDigest"],
        "idempotencyKey": attribution["idempotencyKey"],
        "issuedAt": attribution["issuedAt"],
        "expiresAt": attribution["expiresAt"],
        "issuerKeyId": attribution["issuerKeyId"],
        "signature": {"alg": attribution["signature"]["alg"]},
    }
    return json.dumps(
        covered,
        ensure_ascii=False,
        allow_nan=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")


def attribution_digest(attribution):
    return hashlib.sha256(canonical_signing_bytes(attribution)).hexdigest()


def decode_public_key(value):
    return _canonical_base64url(value, 32, "public key")


def parse_review_reference(value):
    if not isinstance(value, str) or value.count("@sha256:") != 1:
        _error("invalid_review_reference", "expected grv_<id>@sha256:<digest>")
    review_id, publication_digest = value.split("@sha256:")
    if REVIEW_ID.fullmatch(review_id) is None or DIGEST.fullmatch(publication_digest) is None:
        _error("invalid_review_reference", "review id or publication digest is malformed")
    return {"reviewId": review_id, "publicationDigest": publication_digest}


def format_review_reference(review_id, publication_digest):
    value = f"{review_id}@sha256:{publication_digest}"
    parse_review_reference(value)
    return value
